### DISCUSSION TITLE
Manual conditional effects plot with `newdata` and `nsim`

### DISCUSSION THREAD

**Discussion by maxlindmark** [sdmTMB/sdmTMB#161] [Q&A] [answered]
I'm playing around with delta-beta models by fitting two components separately, following the vignette. 

Here's an example of the Binomial part:

```
pcod_mesh <- make_mesh(pcod, c("X", "Y"), cutoff = 15)

fit_dg <- sdmTMB(present ~ depth,
                 data = pcod,
                 mesh = pcod_mesh,
                 time = "year",
                 spatial = "off", 
                 spatiotemporal = "off",
                 family = binomial(link = "logit")
)
```

I want to visualise conditional effects of depth in this case. I gather I should use the `nsim` argument to get uncertainty on the predictions. Since I only want the conditional effect of depth (and also because I have spatial effects turned off in this case), I provide a newdata with a specific year and varying depth. 

```
nd <- data.frame(
  depth = seq(min(pcod$depth),
              max(pcod$depth),
              length.out = 5
  ),
  year = 2015L # a chosen year
)

p <- predict(fit_dg, newdata = nd, nsim = 5)
```

When I predict I get the following warning:

```
Warning message:
The time elements in `newdata` are not identical to those in the original dataset.
This is normally fine, but may create problems for index standardization. 
```

This seems ok, only a warning right? But then I look at the output, it looks like this:

> p
            [,1]       [,2]        [,3]        [,4]        [,5]
2015  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2015 -0.09173057 -0.1385735 -0.06271809 -0.05486111 -0.06765072
2015 -1.34721763 -1.3541357 -1.15869983 -1.19805640 -0.97783094
2015 -2.60270470 -2.5696979 -2.25468156 -2.34125170 -1.88801117
2015 -3.85819176 -3.7852600 -3.35066329 -3.48444699 -2.79819139
2003  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2004  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2005  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2007  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2009  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2011  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2013  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951
2017  1.16375650  1.0769886  1.03326364  1.08833419  0.84252951

I.e., the years in the fitting data **not** provided in `newdata` have been appended to the prediction output. I don't really understand where these come from, because I read this from `?predict.sdmTMB`: _Returns a matrix of nrow(data) by nsim representing the estimates of the linear predictor (i.e., in link space)._

For plotting the sims against a covariate I could filter rows in this matrix that corresponds to `nd`, so that I can join in the depth values from `nd` into the sims output. But at this point I'm a bit hesitant this is the right way to do this, and I wonder if sims is mainly intended for index standardisation and/or gridded predictions (not non-spatial models). Or if I'm misunderstanding something about how `newdata` and `nsim` interact.

How would you approach plotting conditional effects with uncertainty with a delta model that isn't built in?

Thanks for tips and help!

**Comment by maxlindmark** [comment]
Just to clarify, I'm interested in the total delta model prediction eventually (if I only wanted the conditional effect from the Binomial model I could have done it the normal way, but my issue for the total model prediction can be exemplified with only one of the components so that's why I'm doing that)

**Comment by seananderson** [comment, marked as answer]
There are two issues here.

Predicting with missing years is done by adding fake years and removing
them in the end, but I forgot to remove them for the nsim output option.

There’s no reason why we can’t return standard errors on the combined
prediction. We can add that.

One caveat is that the response curve shape depends on the value of the
other predictors when you have two different links.

**Reply by maxlindmark** [reply]
Ah, thanks for clarifying!

I'm not sure I understand the last point though. For instance, what does it mean for the delta_gamma model (with logit and log links) in the visreg vignette? If the model hade included multiple years, would the curve look different for each year even without any interaction?
